torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -44,17 +44,19 @@ bool mergeTypes(ArrayRef<Value*> lhs, ArrayRef<Value*> rhs, ArrayRef<Value*> out
   JIT_ASSERT(lhs.size() == rhs.size() && rhs.size() == outputs.size());
   bool changed = false;
   for(size_t i = 0; i < lhs.size(); ++i) {
+    auto old_output_type = outputs[i]->type();
     if(*lhs[i]->type() == *rhs[i]->type()) {
       outputs[i]->setType(lhs[i]->type());
     } else {
       outputs[i]->setType(DynamicType::get());
-      changed = true;
     }
+    if(*old_output_type != *outputs[i]->type())
+      changed = true;
```

Comment:
Can you please show me an example where this actually does matter? The code is better now, I'm just curious

Reply:
test_shape_analysis_loop causes an infinite loop without this change. I added a note. What happens is that the c, b outputs always have size 5, so it was always thinking the sizes didn't match because it would be merging with a  dynamic size. The change makes it so that we are actually looking at whether we changed something that would be in input to the iteration.